### ISSUE TITLE
NUnit3 agent hangs after encountering an "CannotUnloadAppDomainException"

### ISSUE THREAD

**Issue by linos68** [nunit/nunit-console#259]
I do believe that the nunit-agent process hangs after it run into a problem with unloading app domains after all tests have run from a certain test assembly. Unfortunately we suffer from a small number of "shaky" tests which can lead to the problem in unloading app domains (thread which cannot be aborted or something like this). Of course we are trying to fix these problems in our tests, but nevertheless I think that the test runner never must break because of errors in the tests assemblies.

We're using Nunit3Console version 3.6.1

### Stack trace of the monitored app domain issue taken from TeamCity (version 2017.1) :
_Error : C:\TcTemp\buildTmp\EUSULw5pVgYsE72Ff9xROS7C6Rb9b4jB.nunit
Exception encountered unloading AppDomain

Server stack trace: 
bei NUnit.Engine.Services.DomainManager.DomainUnloader.Unload()
bei NUnit.Engine.Runners.TestDomainRunner.UnloadPackage()
bei NUnit.Engine.Runners.AbstractTestRunner.Dispose(Boolean disposing)
bei NUnit.Engine.Runners.AbstractTestRunner.Dispose()
bei NUnit.Engine.Runners.AggregatingTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
bei System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
bei System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

After this issue has been encountered, the nunit-agent.exe process keeps the test assembly locked. And with it all system resources allocated by the last executed test such as tcp ports. With the effect, that future tests cannot be run successful on the same TeamCity agent again.

STEPS NEEDED

1. Write a test which launches a thread which refuses to stop and abort (Trick here: catch tread abort exception and keep thread busy in a finally block)
2. Run test with TeamCity, NUnit3 with nunit3.console runner 3.6 (Probably TeamCity is not really required, but I didn't check this out so far)

STATUS
The test will fail - as expected. But you will find that the nunit-agent.exe process is still running (check with windows task manager), locking the test assembly. You won't be able to delete it.

**Comment by jnm2** [comment]
@nunit/engine-team Do we already have a facility for surfacing this kind of error where we don't have a specific test to blame?

**Comment by linos68** [comment]
I made some experiments with custom builds of the console runner. Instead of re-throwing exceptions in the "UnloadPackage" method in the process runners (subclasses from AbstractTestRunner) I raised a log.
This seemed to help so far, the nunit-agent.exe process exits as expected.

In the latest experiment I even raised an event on the ITestEventListener. This way these kind of exceptions will also show up in the TeamCity's logs. Otherwise developers won't have a clue why tests from a certain test assembly didn't run. I think that something similar should be done when creating/initializing in case an unexpected exception occurs.

**Comment by kdubau** [comment]
I think I am having the same issue. Any thoughts on how to determine what in my test is causing the issues?

**Comment by ChrisMaddock** [comment]
@kdubau There were a few ideas in the post below, that would require fairly simple modification to the NUnit Console codebase. 

https://github.com/nunit/nunit-console/issues/111#issuecomment-256897931

**Comment by kwojciechowski** [comment]
I have the same issue

**Comment by ChrisMaddock** [comment]
@linos68 @kdubau @kwojciechowski 

Would someone be able to provide a minimal repo which causes the agent to hang? I'm able to reproduce the crash, but the agent is still shutting down cleanly in my repro.

**Comment by ChrisMaddock** [comment]
Theory: this issue will only be seen when running multiple assemblies, without parallelisation. That would be either by specifying that setup manually, or running an .nunit project with multiple assemblies, and without overriding any of the defaults.

Agents should always be terminated properly, regardless of AppDomain unload. However, in the above situation, there is an issue where the Agent shut down code may not be called. That should be resolved by https://github.com/nunit/nunit-console/pull/321.

**Comment by linos68** [comment]
Hi,

I am no longer involved in the project where these issues occured.

But you can force a «CannotUnloadAppDomainException» if you create a thread in a test which refuses to terminate. To do that you need a thread catching all exceptions and stay in a finally block.

I guess there might exist other reasons for such an exception. But with this procedure you’re able to reproduce at least one case.

Cheers
Marco

________________________________
Von: Chris Maddock <notifications@github.com>
Gesendet: Saturday, December 2, 2017 8:19:51 PM
An: nunit/nunit-console
Cc: linos68; Mention
Betreff: Re: [nunit/nunit-console] NUnit3 agent hangs after encountering an "CannotUnloadAppDomainException" (#259)


Theory: this issue will only be seen when running multiple assemblies, without parallelisation. That would be either by specifying that setup manually, or running an .nunit project with multiple assemblies, and without overriding any of the defaults.

Agents should always be terminated properly, regardless of AppDomain unload. However, in the above situation, there is an issue where the Agent shut down code may not be called. That should be resolved by #321<https://github.com/nunit/nunit-console/pull/321>.

**Comment by ChrisMaddock** [comment]
Thanks @linos68. I can repro the exception fine, and the crash, but I haven't been able to repro the agent hanging. In all my attempts, the nunit-agent program shuts down itself.

Here's my repro class.
```
    [TestFixture]
    public class TestClass
    {
        [Test]
        public void TestMethod()
        {
            Task.Factory.StartNew(Bang);
        }

        public void Bang()
        {
            try
            {
                throw new Exception();
            }
            finally
            {
                while(true)
                    Thread.Sleep(1);
            }
        }
```
This triggers the exception. If you put this assembly inside a `.nunit` project with >1 other project, then you get the crash.

But I can't make the nunit-agent.exe hang - it always exits when the console crashes for me. Any ideas on that one welcome! 😄

**Comment by linos68** [comment]
Try this:

        [TestMethod]
        public void Test()
        {
            Thread thread = new Thread(ThreadMain);
            thread.Start();
        }

        private void ThreadMain()
        {
            try
            {
                while (true)
                {
                    Thread.Sleep(1000);
                }
            }
            catch (ThreadAbortException e)  // Attempts to unload app domain will try aborting all threads in the app domain
            {
                Thread.ResetAbort();
            }
            finally
            {
                while (true)
                {
                    Thread.Sleep(1000);
                }
            }
        }

________________________________
Von: Chris Maddock <notifications@github.com>
Gesendet: Saturday, December 2, 2017 10:28:01 PM
An: nunit/nunit-console
Cc: linos68; Mention
Betreff: Re: [nunit/nunit-console] NUnit3 agent hangs after encountering an "CannotUnloadAppDomainException" (#259)


Thanks @linos68<https://github.com/linos68>. I can repro the exception fine, and the crash, but I haven't been able to repro the agent hanging. In all my attempts, the nunit-agent program shuts down itself.

Here's my repro class.

    [TestFixture]
    public class TestClass
    {
        [Test]
        public void TestMethod()
        {
            Task.Factory.StartNew(Bang);
        }

        public void Bang()
        {
            try
            {
                throw new Exception();
            }
            finally
            {
                while(true)
                    Thread.Sleep(1);
            }
        }


This triggers the exception. If you put this assembly inside a .nunit project with >1 other project, then you get the crash.

But I can't make the nunit-agent.exe hang - it always exits when the console crashes for me. Any ideas on that one welcome! 😄

**Comment by ChrisMaddock** [comment]
That did it - thanks. Can confirm this case isn't fixed by #321 - will have a dig.
